whylog/config/rule.py

```diff
@@ -57,6 +65,265 @@ def get_causes_parsers(self):
+        are connected by delta constraint. This edge is directed from base to depended parser.
+        This implementation assume that connectivity parsers graph is a tree.
+        Algorithm steps:
+            1. Calculate effect parser search range. Take primary key values from effect_clues.
+            2. For every parser in rule find its all delta constraints where is a base parser
+                (_aggregate_constraints method). It is connectivity parsers graph creator.
+            3. Do a BFS order over connectivity parsers graph. Start from effect parser which is a root in tree.
+                Calculate parser search range for every child parser of actual parser if it is possible.
+            4. Set maximal search range for all unreachable parser from effect parser in graph
+            5. Pop effect parser search range.
+        """
+        parser_ranges = {
+            self.EFFECT_NUMBER: self._get_effect_range(effect_clues, group, group_type)
+        }
+        queue = deque([self.EFFECT_NUMBER])
+        aggregated_constraints = self._aggregate_constraints()
+        used_parsers = set([self.EFFECT_NUMBER])
+        while queue:
+            parser_number = queue.popleft()
+            for constraint in aggregated_constraints[parser_number]:
+                clues_groups = constraint['clues_groups']
+                depended_parser_number = clues_groups[0][0]
+                base_parser_number = clues_groups[1][0]
+                if depended_parser_number in used_parsers:
+                    continue
+                if not self._is_primary_key_constraint(clues_groups):
+                    continue
+                _, group_type = self._causes[depended_parser_number - 1].get_primary_key_group()
+                parser_ranges[depended_parser_number] = self._calculate_parser_bounds(
+                    base_parser_number, constraint['params'], group_type, parser_ranges
+                )
+                queue.append(depended_parser_number)
+                used_parsers.add(depended_parser_number)
+        self.create_ranges_for_unconnected_parsers(effect_clues, group, group_type, parser_ranges)
+        parser_ranges.pop(self.EFFECT_NUMBER)
+        return parser_ranges
+
+    def _get_effect_range(self, effect_clues, group_number, group_type):
+        """
+        This method basing on effect clues and effect primary key creates effect
+        search range for effect primary key type which in LEFT_BOUND = RIGHT_BOUND and equals
+        primary key value from effect clues.
+        This method is invokes when algorithm with calculate parsers ranges starts. It give
+        set up for this algorithm.
+        Example:
+            clues_groups = {'effect' : Clue((11, 'aaa'), '11 aaa', LineSource('foo', 'bar'))}
+            group_number = 1
+            group_type = 'int'
+            expected returned value : {EFFECT_NUMBER: {'int': {LEFT_BOUND: 11, RIGHT_BOUND: 11}}}
```

Comment:
This dictionary can't be returned by this function. I'm not sure if it is a bug in code or just mistake in docstring. Please resolve this problem.